File: shared/mux_protocol.py
# shared/mux_protocol.py

SYNC1 = 0xAA
SYNC2 = 0x55

CHAN_CRSF = 0x01
CHAN_MAVLINK = 0x02
CHAN_CONFIG = 0x03

class MuxParser:
    def __init__(self):
        self.state = 0 # 0: SYNC1, 1: SYNC2, 2: CHAN_ID, 3: LEN, 4: PAYLOAD, 5: CHECKSUM
        self.chan_id = 0
        self.length = 0
        self.payload = bytearray()
        self.checksum = 0

    def parse_byte(self, b):
        # Returns (success, chan_id, payload)
        if self.state == 0:
            if b == SYNC1:
                self.state = 1
        elif self.state == 1:
            if b == SYNC2:
                self.state = 2
            elif b == SYNC1:
                self.state = 1
            else:
                self.state = 0
        elif self.state == 2:
            if b in [CHAN_CRSF, CHAN_MAVLINK, CHAN_CONFIG]:
                self.chan_id = b
                self.state = 3
            elif b == SYNC1:
                self.state = 1
            else:
                self.state = 0
        elif self.state == 3:
            self.length = b
            self.payload = bytearray()
            if b == 0:
                self.state = 5
            else:
                self.state = 4
        elif self.state == 4:
            self.payload.append(b)
            if len(self.payload) >= self.length:
                self.state = 5
        elif self.state == 5:
            self.checksum = b
            self.state = 0
            calc = calculate_checksum(self.chan_id, self.payload)
            if calc == self.checksum:
                return True, self.chan_id, bytes(self.payload)
        return False, 0, b""

def calculate_checksum(chan_id, payload):
    cksum = (chan_id + len(payload)) & 0xFF
    for b in payload:
        cksum = (cksum + b) & 0xFF
    return cksum

def mux_encode(chan_id, payload):
    if isinstance(payload, str):
        payload = payload.encode('utf-8')
    out = bytearray([SYNC1, SYNC2, chan_id, len(payload)])
    out.extend(payload)
    out.append(calculate_checksum(chan_id, payload))
    return bytes(out)

File: shared/mux_protocol.c
#include "mux_protocol.h"
#include <string.h>

void mux_parser_init(MuxParser *parser) {
    if (!parser) return;
    parser->state = MUX_STATE_SYNC1;
    parser->payload_index = 0;
    parser->calc_checksum = 0;
    memset(&parser->frame, 0, sizeof(MuxFrame));
}

uint8_t mux_calculate_checksum(uint8_t chan_id, uint8_t len, const uint8_t *payload) {
    uint8_t cksum = chan_id + len;
    for (uint8_t i = 0; i < len; i++) {
        cksum += payload[i];
    }
    return cksum;
}

bool mux_parse_byte(MuxParser *parser, uint8_t byte, MuxFrame *out_frame) {
    if (!parser || !out_frame) return false;

    switch (parser->state) {
        case MUX_STATE_SYNC1:
            if (byte == MUX_SYNC1) {
                parser->state = MUX_STATE_SYNC2;
            }
            break;

        case MUX_STATE_SYNC2:
            if (byte == MUX_SYNC2) {
                parser->state = MUX_STATE_CHAN_ID;
            } else if (byte == MUX_SYNC1) {
                // Stay in SYNC2 state if we get another SYNC1
                parser->state = MUX_STATE_SYNC2;
            } else {
                parser->state = MUX_STATE_SYNC1;
            }
            break;

        case MUX_STATE_CHAN_ID:
            if (byte == MUX_CHAN_CRSF || byte == MUX_CHAN_MAVLINK || byte == MUX_CHAN_CONFIG) {
                parser->frame.chan_id = byte;
                parser->state = MUX_STATE_LEN;
            } else if (byte == MUX_SYNC1) {
                parser->state = MUX_STATE_SYNC2;
            } else {
                parser->state = MUX_STATE_SYNC1;
            }
            break;

        case MUX_STATE_LEN:
            parser->frame.len = byte;
            parser->payload_index = 0;
            if (byte == 0) {
                parser->state = MUX_STATE_CHECKSUM;
            } else {
                parser->state = MUX_STATE_PAYLOAD;
            }
            break;

        case MUX_STATE_PAYLOAD:
            parser->frame.payload[parser->payload_index++] = byte;
            if (parser->payload_index >= parser->frame.len || parser->payload_index >= MUX_MAX_PAYLOAD) {
                parser->state = MUX_STATE_CHECKSUM;
            }
            break;

        case MUX_STATE_CHECKSUM:
            parser->frame.checksum = byte;
            parser->calc_checksum = mux_calculate_checksum(parser->frame.chan_id, parser->frame.len, parser->frame.payload);
            parser->state = MUX_STATE_SYNC1; // Always reset to sync state

            if (parser->calc_checksum == parser->frame.checksum) {
                memcpy(out_frame, &parser->frame, sizeof(MuxFrame));
                return true;
            }
            break;
    }

    return false;
}

uint16_t mux_encode(uint8_t chan_id, const uint8_t *payload, uint8_t len, uint8_t *out_buffer) {
    if (!out_buffer) return 0;

    out_buffer[0] = MUX_SYNC1;
    out_buffer[1] = MUX_SYNC2;
    out_buffer[2] = chan_id;
    out_buffer[3] = len;

    if (payload && len > 0) {
        memcpy(&out_buffer[4], payload, len);
    }

    uint8_t cksum = mux_calculate_checksum(chan_id, len, payload);
    out_buffer[4 + len] = cksum;

    return 4 + len + 1;
}
